src/codemagic/universal_apk_generator.py

```diff
@@ -0,0 +1,166 @@
+        argparse_kwargs={'required': False, 'default': None},
+    )
+
+
+SigningInfo = collections.namedtuple('SigningInfo', 'store_path store_pass key_alias key_pass')
+
+
+@cli.common_arguments(
+    UniversalApkGeneratorArgument.PATTERN,
+    UniversalApkGeneratorArgument.BUNDLETOOL_PATH,
+    UniversalApkGeneratorArgument.KEYSTORE_PATH,
+    UniversalApkGeneratorArgument.KEYSTORE_PASSWORD,
+    UniversalApkGeneratorArgument.KEY_ALIAS,
+    UniversalApkGeneratorArgument.KEY_PASSWORD
+)
+class UniversalApkGenerator(cli.CliApp):
+    """
+    Generate universal APK files from Android App Bundles
+    """
+
+    def __init__(self, bundletool_path: pathlib.Path, pattern: pathlib.Path, signing_info: Optional[SigningInfo] = None):
+        super().__init__()
+        self.bundletool_path = bundletool_path
+        self.pattern = pattern
+        self.signing_info = signing_info
+
+    @classmethod
+    def from_cli_args(cls, cli_args: argparse.Namespace):
+        keystore_arguments = (
+            UniversalApkGeneratorArgument.KEYSTORE_PATH,
+            UniversalApkGeneratorArgument.KEYSTORE_PASSWORD,
+            UniversalApkGeneratorArgument.KEY_ALIAS,
+            UniversalApkGeneratorArgument.KEY_PASSWORD,
+        )
+        signing_info_args = [getattr(cli_args, a.value.key) for a in keystore_arguments]
+        if any(signing_info_args) and not all(signing_info_args):
+            raise UniversalApkGeneratorError(
+                'either all signing info arguments should be specified, or none of them should')
+
+        pattern = cli_args.pattern.expanduser()
+        if pattern.is_absolute():
+            pattern = pattern.relative_to(pattern.anchor)
+
+        return UniversalApkGenerator(
+            bundletool_path=cli_args.bundletool_path,
+            pattern=pattern,
+            signing_info=SigningInfo(*signing_info_args) if all(signing_info_args) else None,
+        )
+
+    @cli.action('generate')
+    def generate(self) -> NoReturn:
+        """
+        Generate universal APK files from Android App Bundles
+        """
+        self.logger.info(f'Searching for files in {self.pattern.resolve()}')
+        with tempfile.TemporaryDirectory() as d:
+            if not self.bundletool_path.exists():
+                self.bundletool_path = pathlib.Path(d) / 'bundletool.jar'
+                self.logger.info(f'Downloading bundletool to {self.bundletool_path}')
+                urllib.request.urlretrieve(models.Bundletool.DOWNLOAD_URL, self.bundletool_path)
+
+            did_find_paths = False
+            for path in pathlib.Path().glob(str(self.pattern)):
+                did_find_paths = True
+                self.logger.info(f'Generating universal apk for {path}')
+                apk_path = self._generate_apk(path)
+                self.logger.info(f'Generated {apk_path.resolve()}')
+
+        if not did_find_paths:
+            raise UniversalApkGeneratorError('Did not find any matching files to generate apk from')
```

Comment:
Error message is not complete. From _where_ it didn't find any matching files?

Reply:
I thought it vas valid english, but it might be a germanism. Reworded, but did not add the path because it is mentioned in the first log entry

Reply:
Now that I read it again, it seems very logical indeed. But yesterday, for whatever reason, I managed to convince myself strongly that it should state where it was looking the bundles from.